src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java

```diff
@@ -0,0 +1,169 @@
+                .bcat(CollectionUtils.isNotEmpty(blockedAdvCat) ? blockedAdvCat : request.getBcat())
+                .imp(Collections.singletonList(updatedImp))
+                .site(newSite)
+                .build();
+    }
+
+    private ExtImpTaboola parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TABOOLA_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException("Missing bidder ext in impression with id: " + imp.getId());
+        }
+    }
+
+    private String buildEndpointUrl(ExtImpTaboola extImpTaboola, String type, String domain) {
+        return endpointTemplate.replace("{{Host}}", domain)
+                .replace("{{MediaType}}", type)
+                .replace("{{PublisherID}}", extImpTaboola.getPublisherId());
+    }
+
+    private static void validateImp(Imp imp) {
+        if (imp.getBanner() == null && imp.getXNative() == null) {
+            throw new PreBidException("For Imp ID %s Banner or Native is undefined".formatted(imp.getId()));
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, ExtImpTaboola impExt) {
+        return imp.toBuilder()
+                .tagid(impExt.getPublisherId())
+                .bidfloor(impExt.getBidFloor())
+                .build();
+    }
+
+    private static BidType getBidType(Imp imp) {
+        if (imp.getBanner() != null) {
+            return BidType.banner;
+        } else if (imp.getXNative() != null) {
+            return BidType.xNative;
+        } else {
+            throw new PreBidException("Failed to find impression for ID: " + imp.getId());
+        }
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.withValues(extractBids(httpCall.getRequest().getPayload(), bidResponse));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private List<BidderBid> bidsFromResponse(BidRequest bidRequest,
+                                             BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, getBidType(bidRequest.getImp().get(0)), bidResponse.getCur()))
```

Comment:
In GO they used all of imps to recognize `BidType`.

Reply:
Also, if there is some problems, they just skip that bid. Do not skip all of bids on first error.